flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -166,21 +166,40 @@ struct CharacterOpsBuilderImpl {
       return fir::ReferenceType::get(getCharacterType());
     }
 
-    bool isConstant() const { return data.getType().isa<fir::SequenceType>(); }
+    bool isNotInMemory() const {
+      return data.getType().isa<fir::SequenceType>() ||
+             data.getType().isa<fir::CharacterType>();
+    }
+
+    std::optional<fir::SequenceType::Extent> getCompileTimeLenght() const {
+      auto type = data.getType();
+      if (type.isa<fir::CharacterType>())
+        return 1;
+      if (auto refType = type.dyn_cast<fir::ReferenceType>())
+        type = refType.getEleTy();
+      if (auto seqType = type.dyn_cast<fir::SequenceType>()) {
+        auto shape = seqType.getShape();
+        assert(shape.size() == 1 && "only scalar character supported");
+        return shape[0];
+      }
+      return {};
```

Comment:
Do we need to handle adjusted sizes?
```
CHARACTER(LEN=n) :: var
```
?

Reply:
You mean, if "n" is a constant ? That would be nice, right now it's not as easy as the others because then "var" comes here as a boxchar right know (we could look at the length producer op, but I am not sure their is a precedent doing so in the bridge an I am not sure it is something we should start doing).
However, in the local variable repackaging, we could maybe use `fir.array<n x fir.char<kind>>` in the symbol map instead of `fir.boxchar<kind>` if the adjustment is constant.

Reply:
I meant `n` as a variable. The length of `var` being unknown at compile-time.

Reply:
We should lower
```fortran
SUBROUTINE s(n)
...
CHARACTER(LEN=n) :: var
```
to something like
```mlir
%var = fir.alloca !fir.array<?x!fir.char<1>>, %arg0 { name = "var" }
```
The size from the SequenceType will be unknown, but the size is known at runtime. It may not be a concern here.

Reply:
OK, then this code would just return nullopt here and we would generate the 3 loops character  assignment here in all cases.
The idea of using the knowledge of the compile time length is just to avoid generating bunches of loop when we know from the start they are not needed (especially for lenght one assignment that should yield minimal code from the start). I know later optimization could get rid of this, currently that is not the case right now.

Regarding the type of var, it goes in the discussion of how we have to update expression lowering (and the FIR generations utilities) to get the length and shape information when needed (Right now the character utilities would die on `!fir.ref<!fir.array<?x!fir.char<1>>>` because the character utility had no way to get the dynamic length from that.